### PR TITLE
duo implementation for pingfed, with help from shibboleth's impl

### PR DESCRIPTION
**Pull request by willysamz**
#282

### CODE REVIEW

**Comment by munkyboy** [comment]
just tested it and it works, thanks!

**Comment by charlieb00** [comment]
This also worked well for me.  What are the chances of getting this merged?

**Comment by jpmontez** [comment]
Would love to see this merged. My organization uses Duo and I'm out of luck with using the CLI as it stands today.

**Comment by hasit** [comment]
We use Duo as well. Looking forward to get this merged.

**Comment by ntroast** [comment]
@wolfeidau review please?

File: pkg/provider/pingfed/pingfed.go [hunk 1]
```diff
@@ -226,10 +477,48 @@ func extractSAMLResponse(doc *goquery.Document) (v string, ok bool) {
...
+
+func parseTokens(blob string) (string, string, string, string) {
+
+	hostRgx := regexp.MustCompile(`'host': '(.*?)'`)
+	sigRgx := regexp.MustCompile(`'sig_request': '(.*?)'`)
+	dpaRgx := regexp.MustCompile(`'post_action': '(.*?)'`)
+	dataSigRequest := sigRgx.FindStringSubmatch(blob)
+	duoHost := hostRgx.FindStringSubmatch(blob)
+	postAction := dpaRgx.FindStringSubmatch(blob)
+	duoSignatures := strings.Split(dataSigRequest[1], ":")
```
Comment (wolfeidau) [inline pkg/provider/pingfed/pingfed.go]: I would check the array has assumed number of values before accessing an index.

Comment (willysamz) [reply pkg/provider/pingfed/pingfed.go]: Added check for assumed length before accessing

**Comment by willysamz** [comment]
@wolfeidau anything else you need from me?

**Comment by ntroast** [comment]
@wolfeidau re-review?

**Comment by sandinak** [comment]
Any chance we can get this merged?

**Comment by rmtzcx** [comment]
Looking forward to get this merged.

**Review by mapkon — CHANGES_REQUESTED** [review]
House keeping alert: I am going to close this PR if I do not get traction on it in the next days.

**Pull request closed without merging**